Update `eks/cluster` README (#616)
cloudposse/terraform-aws-components: modules/eks/cluster/README.md

# Component: `eks/cluster`

This component is responsible for provisioning an end-to-end EKS Cluster, including managed node groups.

> :::info

Reviewer: Maybe should be a warning?

> This component can only be deployed after logging in to AWS via Federated login with SAML (e.g. GSuite) or assuming an IAM role (e.g. from a CI/CD system). It cannot be deployed if you login to AWS via AWS SSO, the reason being is that on initial deployment, the EKS cluster will be owned by the assumed role that provisioned it. If this were to be the AWS SSO Role, then we risk losing access to the EKS cluster once the ARN of the AWS SSO Role eventually changes.

Author: ```suggestion
This component can only be deployed after logging into AWS via Federated login with SAML (e.g. GSuite) or assuming an IAM role (e.g. from a CI/CD system). It cannot be deployed if you log into AWS via AWS SSO, the reason being that on initial deployment, the EKS cluster will be owned by the assumed role that provisioned it. If this were to be the AWS SSO Role, then we risk losing access to the EKS cluster once the ARN of the AWS SSO Role eventually changes.
```

:::

## Usage

> **Stack Level**: Regional

Reviewer: Do you think we should keep this?

Author: Yes because some customers only use SSO. This is one of the few (only?) spaces we require SAML

              - system:masters
            role: spacelift
        # Roles from the account owning the cluster to allow access to the cluster
        # See `aws-team-roles` component.
        delegated_iam_roles:
          - groups:
              - system:masters
              - idp:ops
            role: admin
          - groups:
              - idp:poweruser
            role: poweruser
          - groups:
              - idp:observer
            role: observer
          - groups:
              - system:masters
            role: terraform
        # Roles from AWS SSO allowing cluster access
        # See `aws-sso` component.
        sso_iam_roles:
          - groups:
              - idp:observer
              - idp:observer-extra
            role: ReadOnlyAccess
          - groups:
              - system:masters
              - idp:ops
            role: AdministratorAccess
        fargate_profiles:
          karpenter:
            kubernetes_namespace: karpenter
            kubernetes_labels: null
        karpenter_iam_role_enabled: true
```

### Usage with Node Groups

The `eks/cluster` component also supports node groups! In order to add a set list of nodes to 

> provision with the cluster, add values for `var.managed_node_groups_enabled` and `var.node_groups`.

Reviewer: Can node groups and karpenter be deployed at the same time? If so, please mention that.

Author: added note

# Component: `eks/cluster`

This component is responsible for provisioning an end-to-end EKS Cluster, including managed node groups.

:::warning

> This component can only be deployed after logging into AWS via Federated login with SAML (e.g. GSuite) or assuming an IAM role (e.g. from a CI/CD system). It cannot be deployed if you log into AWS via AWS SSO, the reason being that on initial deployment, the EKS cluster will be owned by the assumed role that provisioned it. If this were to be the AWS SSO Role, then we risk losing access to the EKS cluster once the ARN of the AWS SSO Role eventually changes.

Reviewer: ```suggestion
This component should only be deployed after logging into AWS via Federated login with SAML (e.g. GSuite) or assuming an IAM role (e.g. from a CI/CD system). It should not be deployed if you log into AWS via AWS SSO, the reason being that on initial deployment, the EKS cluster will be owned by the assumed role that provisioned it. If this were to be the AWS SSO Role, then we risk losing access to the EKS cluster once the ARN of the AWS SSO Role eventually changes.
```

              - system:masters
              - idp:ops
            role: admin
          - groups:
              - idp:poweruser
            role: poweruser
          - groups:
              - idp:observer
            role: observer
          - groups:
              - system:masters
            role: terraform
        # Roles from AWS SSO allowing cluster access
        # See `aws-sso` component.
        sso_iam_roles:
          - groups:
              - idp:observer
              - idp:observer-extra
            role: ReadOnlyAccess
          - groups:
              - system:masters
              - idp:ops
            role: AdministratorAccess
        fargate_profiles:
          karpenter:
            kubernetes_namespace: karpenter
            kubernetes_labels: null
        karpenter_iam_role_enabled: true
```

### Usage with Node Groups

The `eks/cluster` component also supports node groups! In order to add a set list of nodes to 
provision with the cluster, add values for `var.managed_node_groups_enabled` and `var.node_groups`.

:::info
You can use manage node groups in conjunction with Karpenter node groups! Simply deploy node groups as demonstrated below,

> and then deploy Karpenter with the necessary components.
> 

Reviewer: ```suggestion
and then deploy Karpenter with the necessary components.
```

Author: no change?

Reviewer: Just deleting the new line

```yaml
components:
  terraform:
    eks/cluster:
      vars:
        enabled: true
        name: eks
        iam_primary_roles_tenant_name: core
        cluster_kubernetes_version: "1.25"
        availability_zones: ["us-east-1a", "us-east-1b", "us-east-1c"]
        aws_ssm_agent_enabled: true
        allow_ingress_from_vpc_accounts:
          - tenant: core
            stage: auto
          - tenant: core
            stage: corp
          - tenant: core
            stage: network
        public_access_cidrs: []
        allowed_cidr_blocks: []
        allowed_security_groups: []
        enabled_cluster_log_types:

>           - api

Reviewer: Let's keep but add comment. This is just an example _in_ the readme.

>           - api

Reviewer: ```suggestion
          # Caution: enabling `api` log events may lead to a substantial increase in Cloudwatch Logs expenses.
          - api

          - authenticator
          - controllerManager
          - scheduler
        oidc_provider_enabled: true
        public_access_cidrs: ["72.107.0.0/24"]
        managed_node_groups_enabled: true
        node_groups: # null means use default set in defaults.auto.tf.vars
          main:
            # values of `null` will be replaced with default values
            # availability_zones = null will create 1 auto scaling group in
            # each availability zone in region_availability_zones
            availability_zones: null

            desired_group_size: 3 # number of instances to start with, must be >= number of AZs
            min_group_size: 3 # must be  >= number of AZs
            max_group_size: 6

            # Can only set one of ami_release_version or kubernetes_version
            # Leave both null to use latest AMI for Cluster Kubernetes version
            kubernetes_version: null # use cluster Kubernetes version
            ami_release_version: null # use latest AMI for Kubernetes version

            attributes: []
            create_before_destroy: true
            disk_size: 100
            cluster_autoscaler_enabled: true
            instance_types:

>               - t3.medium

Reviewer: Shouldn't we still provide an example of how to use node groups? Since node groups are still supported. We need to be able to reference alternative configurations. perhaps a separate example?
